### ISSUE TITLE
Fix data formats propagation for TensorFlow modules

### ISSUE THREAD

**Issue by pilkicTT** [tenstorrent/tt-forge-fe#281]
The data types are not properly propagated from TensorFlow module to the Forge module (they always get overridden as `float32`).

See test `test_conv2d` in `forge/test/test_ops_tf.py`.

**Comment by LPanosTT** [comment]
So in the latest main, TF input activation data types are passed correctly, but not parameters. After a little digging I've found this section of code that is responsible:  line 709 in `third_party/tvm/python/tvm/relay/frontend/tensorflow.py`

```
unsuported_types = [dtypes.bfloat16, ]
np_array = tensor_util.MakeNdarray(value.tensor)
orig_dtype = str(dtypes.as_dtype(value.tensor.dtype))[9:-2] # "i.e. <dtype: 'int32'> -> int32"
if dtypes.as_dtype(value.tensor.dtype) in unsuported_types:
    np_array = np_array.astype('float32')
```

This is here because bfloat16 is not a numpy data type.

**Comment by nvukobratTT** [comment]
> So in the latest main, TF input activation data types are passed correctly, but not parameters. After a little digging I've found this section of code that is responsible: line 709 in `third_party/tvm/python/tvm/relay/frontend/tensorflow.py`
> 
> ```
> unsuported_types = [dtypes.bfloat16, ]
> np_array = tensor_util.MakeNdarray(value.tensor)
> orig_dtype = str(dtypes.as_dtype(value.tensor.dtype))[9:-2] # "i.e. <dtype: 'int32'> -> int32"
> if dtypes.as_dtype(value.tensor.dtype) in unsuported_types:
>     np_array = np_array.astype('float32')
> ```
> 
> This is done whilst building the TVM relay module from the tensorflow GraphDef.
> 
> This is here because bfloat16 is not a numpy data type.

Good catch @LPanosTT! 

Can we model this as convert it properly to bfloat16 at some point? E.g., if we need to keep it in float32 during TVM compile, let's keep track of it one way or another, and do a proper conversion bit before or during codegen to ForgeModule. 

What do you think? :))

**Comment by LPanosTT** [comment]
> Can we model this as convert it properly to bfloat16 at some point? E.g., if we need to keep it in float32 during TVM compile, let's keep track of it one way or another, and do a proper conversion bit before or during codegen to ForgeModule.

Yes, I'm trying to get this working right now. This isn't an issue for PyTorch, I think its simply the fact that the flow for tensorflow converts the parameters to numpy.

**Comment by LPanosTT** [comment]
@nvukobratTT @pilkicTT. So I can get the input parameters for the TF modules to hold the proper datatype by passing `var.framework_dtype` in line 1119 of `third_party/tvm/python/tvm/relay/frontend/common.py`. However, all the ops have their "dtype" set to float32 still. 

So, we end up generating a module like this

```
class Conv2D(ForgeModule):
    def __init__(self, name):
        super().__init__(name)
        self.add_parameter("conv2d/conv2d_1/kernel:0", forge.Parameter(*(4, 4, 16, 64), requires_grad=True, dev_data_format=forge.DataFormat.Float16_b))

    def forward(self, inputs_0):
        cast_2 = forge.op.Identity("", inputs_0).set_src_layer("conv2d/Cast")
        transpose_3 = forge.op.Transpose("", self.get_parameter("conv2d/conv2d_1/kernel:0"), dim0=-4, dim1=-1, z_dim_slice=64, out_dtype=torch.float32)
        transpose_4 = forge.op.Transpose("", transpose_3, dim0=-3, dim1=-2, z_dim_slice=16, out_dtype=torch.float32)
        transpose_3._value = None
        transpose_5 = forge.op.Transpose("", transpose_4, dim0=-2, dim1=-1, out_dtype=torch.float32)
        transpose_4._value = None
        conv2d_6 = forge.op.Conv2d("", cast_2, transpose_5, stride=[1, 1], padding=[0, 0, 0, 0], dilation=1, groups=1, channel_last=1).set_src_layer("conv2d/conv2d_1/Conv2D")
        cast_2._value = None
        transpose_5._value = None
        return conv2d_6
```

Notice how the `out_dtype` attribute of the transposes is set to `float32`.  These transposes get constevaled into the weight and the output dtype of the weight thus becomes `float32`. 

Not sure why we even have `out_dtype` as an attribute to transpose. If it was something that was only really needed in buda, I suggest we remove it and if the output data type for any op must be forced to something we should insert a separate cast op.

**Comment by nvukobratTT** [comment]
Sounds like a bit larger change, as you'll have to update transpose op through the stack. 

Few details and questions from my side:
1. From where is this cast op coming, and to what it cast? 
2. Does TVM transpose have a specific op dtype? 
3. Ops on TVM side often infer dtype from producer nodes, there might be a case that this cast causes the issues. However, for kernel transpose that shouldn't be the case
4. Maybe we just have a bug how we handle cast op on TVM level. Can you look into more details what is happening there?

**Comment by nvukobratTT** [comment]
Closing as this effort will be tacked as part of the data formats milestone coming in the future.
